src/mpid/ch4/shm/xpmem/xpmem_seg.h

```diff
@@ -0,0 +1,643 @@
+                                                                xpmem_apid_t apid,
+                                                                MPIDI_XPMEM_seg_t ** seg_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_seg_t *seg = NULL;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEGTREE_DO_CREATE_NODE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEGTREE_DO_CREATE_NODE);
+
+    mpi_errno = MPIDI_XPMEM_seg_do_create(seg_ptr, low, high, apid);
+    if (mpi_errno != MPI_SUCCESS)
+        MPIR_ERR_POP(mpi_errno);
+
+    seg = *seg_ptr;
+    seg->height = 1;
+    seg->parent = NULL;
+    seg->left = NULL;
+    seg->right = NULL;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_SEGTREE_DO_CREATE_NODE);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_segtree_do_search_and_insert_safe
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_segtree_do_search_and_insert_safe(MPIDI_XPMEM_segtree_t
+                                                                           * tree, uint64_t low,
+                                                                           uint64_t high,
+                                                                           xpmem_apid_t apid,
+                                                                           MPIDI_XPMEM_seg_t **
+                                                                           seg_ptr, off_t * voffset)
+{
+    int direction;
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_seg_t *sub_root = NULL;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEGTREE_DO_SEARCH_AND_INSERT_SAFE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEGTREE_DO_SEARCH_AND_INSERT_SAFE);
+
+    MPID_THREAD_CS_ENTER(VCI, tree->lock);
+
+    if (tree->root == NULL) {
+        mpi_errno = MPIDI_XPMEM_segtree_do_create_node(low, high, apid, seg_ptr);
+        if (mpi_errno != MPI_SUCCESS)
+            MPIR_ERR_POP(mpi_errno);
+
+        *voffset = 0;
+        tree->root = *seg_ptr;
+        tree->tree_size = 1;
+    } else {
+        sub_root = tree->root;
+
+        MPIDI_XPMEM_AVL_DECLARE_STACK(node_stack, MPIDI_XPMEM_seg_t *, MPIDI_XPMEM_AVL_STACK_SIZE);
+
+        do {
+            if (low < sub_root->low) {
+                /* Insert into left branch */
+                if (sub_root->left != NULL) {
+                    MPIDI_XPMEM_AVL_STACK_PUSH(node_stack, sub_root);
+                    sub_root = sub_root->left;
+                    continue;
+                } else {
+                    direction = AVL_LEFT;
+                }
+            } else if (high <= sub_root->high) {
+                /* Hit the cache */
+                *seg_ptr = sub_root;
+                /* voffset is the offset between registered virtual address in
+                 * cached segment and the virtual address we actually need. */
```

Comment:
This sort of information needs to be in a comment at the beginning of the function documenting all of the input/output parameters so it's easy to find.

Reply:
Improved the comments in xpmem_seg.h